### PR TITLE
Ray.wait num_returns change of behavior

### LINKED ISSUE #6667: ray.wait's num_returns should not fail if num_returns > len(results)
### Describe your feature request

My apologies for the very verbose and syntactic title.
The problem with `ray.wait` happens here - https://github.com/ray-project/ray/blob/480206eef86aba8cf44116de663e044ef8b24cd1/python/ray/worker.py#L1626
It makes no sense to throw for such an 'error', and it effectively forces users to then use e.g. `ray.wait(results, num_returns=min(len(results), N))`.

Instead of raising an exception, a more appropriate solution IMO would be to raise a warning and notify the user the num_returns has been modified to fit to the number of object provided.

As it is right now, users cannot effectively use `num_returns` unless the number of objects provided is divisible by some number. As an example, perhaps users would like to act on 5% of the data at a time, and unless there are 20N number of objects, the users would have to round `num_returns` to the nearest integer. Then `ray.wait` will crash on the final `wait`.

### PR DESCRIPTION
**Pull request by deanwampler**
# Why are these changes needed?

We should be "forgiving" and not throw an exception if the user specifies a `num_returns` that's > `len(ids)`. It's analogous to allowing `str.substring(5, 10000000)` to work a string much shorter than the 2nd argument. 

I also greatly expanded the tests for ray.wait, extracting the previous test with the new tests into a new test file.

## Related issue number

Fixes #6667 

## Checks

- [] I've run `scripts/format.sh` to lint the changes in this PR. 
   I would have, but the `flake8` command fails because of an unrecognized argument `--inline-quotes`.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: CONTRIBUTING.rst
Comment (edoakes) [inline CONTRIBUTING.rst]: thanks for updating - love it!

File: python/ray/tests/test_wait.py
Comment (edoakes) [inline python/ray/tests/test_wait.py]: I think you'll need to add this to `python/ray/tests/BUILD` so that it runs in the CI

File: python/ray/tests/test_wait.py
Comment (edoakes) [inline python/ray/tests/test_wait.py]: This seems to be the only place that we're using a nonzero argument to `f`. Can we force the second ID to not be ready in a different way (e.g., by using`ray.ObjectID.from_random()`) instead and remove the `time.sleep()` call from `f`? Want to discourage sleeping in tests as much as possible as it's a slippery slope to flaky tests.

**Review by edoakes — APPROVED** [review]
LGTM just make sure it's running and passing in the CI

**Comment by edoakes** [comment]
Also looks like the linter is failing:
https://travis-ci.com/ray-project/ray/jobs/282617763

You can run it locally using `ci/travis/format.sh`. If you don't have the right tools installed it should tell you how to install them. You can also run `bash setup_hooks.sh` to set up a git hook that will run this when you push to a remote branch.

File: python/ray/worker.py
Comment (robertnishihara) [inline python/ray/worker.py]: This can be a bit confusing, so here's what I'd suggest. For `num_returns`, document it as

*The maximum number of ready object IDs that should be returned. This value must be at least 1. If it is greater than len(object_ids), it is replaced with len(object_ids). If no timeout is specified, exactly num_returns ready object IDs will be returned. If a timeout is specified, and there are not yet num_returns ready object IDs, then all of the ready object IDs will be returned.*

**Comment by robertnishihara** [comment]
I haven't reviewed the code other than the body of `ray.wait`. However, I have some thoughts on how we should go about API decisions. Will say more in the morning.

**Comment by robertnishihara** [comment]
For now, requesting @pcmoritz's review on the API change only.

**Review by pcmoritz — APPROVED** [review]
The API change looks good to me. At a higher level, I always found wait more confusing to newcomers than needed (if they know UNIX select, they typically get it instantly, otherwise it can be a bit hard to grasp). A simpler API would be to have a function that given a bunch of object ids gives us an iterator over the results in the order they arrive.

File: python/ray/tests/test_wait.py
Comment (deanwampler) [reply python/ray/tests/test_wait.py]: Oh! I assumed (wrongly) that tests are automatically detected. Will fix.

File: python/ray/tests/test_wait.py
Comment (deanwampler) [reply python/ray/tests/test_wait.py]: Pushed a change to use `ray.ObjectID.from_random()`. Appears to work ;)

File: python/ray/worker.py
Comment (deanwampler) [reply python/ray/worker.py]: Changed.

Comment (deanwampler) [reply python/ray/worker.py]: Fixed linter error.

**Comment by deanwampler** [comment]
Pushed an update that merges in `upstream/master` and addresses the comments above.

**Comment by deanwampler** [comment]
Any other feedback or recommended changes?

**Review by zhijunfu — APPROVED** [review]

**Comment by deanwampler** [comment]
Any particular reason this PR hasn't been merged?

**Comment by edoakes** [comment]
@deanwampler lint failing, will merge once fixed:
https://travis-ci.com/ray-project/ray/jobs/283149817

File: doc/source/installation.rst
Comment (richardliaw) [inline doc/source/installation.rst]: why not just `pip install "ray[dashboard]"` in L140?

Comment (robertnishihara) [reply doc/source/installation.rst]: Actually, can we get rid of `ray[dashboard]` now? @ijrsvt @simon-mo

Comment (simon-mo) [reply doc/source/installation.rst]: Yeah `ray[dashboard]` is an empty target now for compatibility reason.

**Comment by deanwampler** [comment]
I'm not sure when it's planned to merge this, but a while ago Robert also suggested that I remove the warning if you provide an int for the timeout rather than a float. I could add that change to this PR or wait until it's merged, then create a new PR. Let me know what you think.

**Comment by robertnishihara** [comment]
@deanwampler I think that's already been done. But if not, it should be a separate PR.

**Review by ericl — COMMENTED** [review]
Lgtm

**Comment by edoakes** [comment]
@deanwampler lint is failing but I think you just need to pull in latest master. Could you do that?

**Comment by deanwampler** [comment]
ping...

**Comment by edoakes** [comment]
@deanwampler lint is still failing:
https://travis-ci.com/ray-project/ray/jobs/294201813

**Comment by deanwampler** [comment]
I see this is still open. I can merge out from master to resolve the conflicts if there is still interest in this change. Otherwise, let's close this PR.

**Comment by robertnishihara** [comment]
Closing for now then.

**Pull request closed without merging**
